### DISCUSSION TITLE
How to import a Rust method which returns data?

### DISCUSSION THREAD

**Discussion by Jackolix** [slint-ui/slint#5333] [Q&A] [answered]
Hey,
im trying to get a String List from a Rust method to create Button elements.
I get an error with the import of the rust method which i dont know how to solve.
**Error: Unsupported foreign import "customers.rs"**

Slint Code: 
``
import {fetch_customers} from "customers.rs";
 searchBox := ListView {
        y: 163px;
        width: 180px;
        height: 150px;
        
        for customer in fetch_customers() : Rectangle {
            height: 40px;
            width: parent.width;
            
            Button {
                x: 0;
                text: data.text;
                //color: white;
                width: parent.width;
            }
        }
        opacity: 0;

        animate y {
            duration: 500ms;
        }
    }
``
    
   Rust Method: 
   
  ``pub fn fetch_customers() -> Vec<String> {
    let client = Client::new();
    let url = "";

    let response = client
        .get(url)
        .send()
        .await
        .expect("Failed to send request");

    let customers: Vec<Customer> = response
        .json()
        .await
        .expect("Failed to parse response as JSON");

    customers
        .into_iter()
        .map(|customer| customer.name)
        .collect()
}
``

Any help is appreciated 🙏 
Thanks

**Comment by hunger** [comment, marked as answer]
You will need to define a `Customer` struct in Slint: 

```slint
struct Customer {
   name: string,
   age: int,
}
```

Then you can have a `in property <[Customer]> model;` in your main slint component with a list of those customers.

You use that model like this:

```slint
    for customer in root.model : Rectangle {
        height: 40px;
        width: parent.width;
        
        Button {
            x: 0;
            text: data.text;
            //color: white;
            width: parent.width;
        }
    }
```

Then you implement that model on the rust side (the Customer stuct is visible there!) and set it into the slint property.

The benefit of using a model is that you can add/remove/change elements in the model and that will do the right things...

In your approach it would get a fresh array of elements each time it needs to render the customers... which transfers a lot of data and will rerender anything from scratch each time. That is hugely inefficient and will also trash your UI state like scroll location and such.

The docs to a `VecModel` on the rust side are here: https://releases.slint.dev/1.6.0/docs/rust/slint/struct.vecmodel

There is an example using a VecModel here: https://github.com/slint-ui/slint/blob/1a9d09be62ed0181dcd349c3e292608e57fdfc48/examples/todo/rust/lib.rs#L27

I hope this helps!

**Reply by Jackolix** [reply]
Thank you!
You helped me alot
